trieste/models/keras/sampler.py

```diff
@@ -80,29 +83,183 @@ def resample_trajectory(self, trajectory: TrajectoryFunction) -> TrajectoryFunct
+    of the stochastic forward passes as a trajectory.
+    """
+
+    def __init__(self, model: TrainableProbabilisticModel):
+        """
+        :param model: The Monte Carlo Dropout model to sample from.
+        """
+        if not isinstance(model, TrainableProbabilisticModel):
+            raise NotImplementedError(
+                f"DropoutTrajectorySampler only works with Probabilistic Models; "
+                f"received {model.__repr__()}"
+            )
+
+        super().__init__(model)
+
+        self._model = model
+
+    def __repr__(self) -> str:
+        """"""
+        return f"{self.__class__.__name__}({self._model!r}"
+
+    def get_trajectory(self) -> TrajectoryFunction:
+        """
+        Generate an approximate function draw (trajectory) by using the predicted means
+    of the stochastic forward passes as a trajectory.
+
+        :return: A trajectory function representing an approximate trajectory
+            from the model, taking an input of shape `[N, 1, D]` and returning shape `[N, 1]`.
+        """
+
+        return dropout_trajectory(self._model)
+
+    def resample_trajectory(self, trajectory: TrajectoryFunction) -> TrajectoryFunction:
+        """
+        Efficiently resample a :const:`TrajectoryFunction` in-place to avoid function retracing
+        with every new sample.
+
+        :param trajectory: The trajectory function to be resampled.
+        :return: The new resampled trajectory function.
+        """
+        tf.debugging.Assert(isinstance(trajectory, dropout_trajectory), [])
+        trajectory.resample()  # type: ignore
+        return trajectory
+
+# @tf.function
+class dropout_trajectory(TrajectoryFunctionClass):
+    """
+    Generate an approximate function draw (trajectory) by using the predicted means
+    of the stochastic forward passes as a trajectory. 
     """
 
-    def __init__(self, model: EnsembleModel):
+    def __init__(self, model: TrainableProbabilisticModel):
         """
         :param model: The model of the objective function.
         """
         self._model = model
-        self._network_index = tf.Variable(self._model.sample_index(1)[0])
-
-    @tf.function
-    def __call__(self, x: TensorType) -> TensorType:  # [N, 1, d] -> [N, 1]
-        """Call trajectory function."""
-        tf.debugging.assert_shapes(
-            [(x, [..., 1, None])],
-            message="This trajectory only supports batch sizes of one.",
+        self._initialized = tf.Variable(False, trainable=False)
+
+        self._batch_size = tf.Variable(
+            0, dtype=tf.int32, trainable=False
         )
-        x = tf.squeeze(x, -2)  # [N, D]
-        return self._model.predict_ensemble(x)[0][self._network_index]
+
+        self._seeds = tf.Variable(
+            tf.ones([0,0], dtype=tf.int32), shape=[None, None], trainable=False
+        )
+            
+    def __call__(self, x: TensorType) -> TensorType:  # [N, B, d] -> [N, B]
```

Comment:
you need tf.function decorator for this method

Reply:
Yes, that's what I was working on in one of the branches, but failed to find a way of making it deterministic. Discussing it with Clinton, he proposed a viable alternative, but it would require some recoding of the layers and elsewhere in the model. I was going to bring it up tomorrow in our morning meeting.